src/ctap2/mod.rs

```diff
@@ -130,6 +144,18 @@ macro_rules! handle_errors {
                 $callback.call(Err(AuthenticatorError::CredentialExcluded));
                 break;
             }
+            HIDError::Command(CommandError::StatusCode(StatusCode::PinAuthInvalid, _))
+                if $cached_puat =>
+            {
+                // We used the cached PUAT, but it was invalid. So we just try again
+                // without the cached one and potentially trigger a new PIN/UV entry from
+                // the user. This happens e.g. in interactive-mode, when we first
+                // executed some bio-related operations with a bio-PUAT and then some
+                // credential management with the bio-PUAT, which fails.
+                $cached_puat = false;
+                $skip_puap = false;
+                continue;
+            }
```

Comment:
Why would that happen? We should be storing permissions with the cached `PinUvAuthToken` and checking them prior to sending any commands. I suppose we might still need this code to handle PUATs that have expired, so maybe only the example in the comment needs to be changed.

Reply:
I actually used it as the comment described. But we should indeed check the permissions before using it.

On a general note: I wanted to have this caching-mechanism as lenient as possible, to never cause non-recoverable errors. If the cached PUAT is bad for whatever reason, auth-rs should just skip it and try again to get a new one. 
Hence the change in `handle_errors!()`. And yes, we also need this for expired ones. But also for "bad" ones, since they are externally provided (all zeros for example, or just a random byte-array).